### PR TITLE
Ingest the checkpoints config topic

### PR DESCRIPTION
**Pull request by facumenzella**
### Motivation
get the `checkpoints` config topic into the SDK so rules are readable before there's anything evaluating them.
https://linear.app/revenuecat/issue/WFL-429/ios-ingest-the-checkpoints-config-topic

Android counterpart: RevenueCat/purchases-android#3895
<details>
<summary>AI session context</summary>

# AI Context

## Metadata

- PR: this one
- Branch: `facu/checkpoints-config-topic`
- Author / human owner: @facumenzella
- Agent(s): Claude Code (Opus 5, 1M context)
- Session source: current conversation
- Generated: 2026-08-05
- Context document version: 1

## Goal

Ingest the `checkpoints` remote-config topic on iOS: add the `RemoteConfigTopic` case, parsing models for the rules schema, and a `CheckpointsConfigProvider` following the `WorkflowsConfigProvider` pattern. Rules parsed only, kept in memory, no evaluation. No user-visible behavior.

## Initial Prompt

"[iOS] Ingest the checkpoints config topic. Add the new topic to the remote-config machinery: `RemoteConfigTopic.Checkpoints("checkpoints")`, parsing models for the schema, and a `CheckpointsConfigProvider` following the `WorkflowsConfigProvider` pattern. Rules only need to be parsed, no evaluation. Keep the parsed rules in memory." (given alongside a link to the Checkpoints product spec)

## Important Follow-up Prompts

- Pointed at the "Storing and serving checkpoint rules" proposal, which supplied the real served schema and replaced an earlier derived guess.
- Asked to compare against purchases-android#3887 and #3895.
- "about 2, we should support both" (inline content and blob), later reversed by Toni's reply.
- "about 1, looks like its checkpoints" (topic wire name).
- Relayed @tonidero's reply that `blobData` already abstracts inlined vs downloaded blobs and `topic()` is only needed for metadata.
- "can you double-check this by checking the other topic? the one for workflows"

## Agent Contribution

- Implemented `RemoteConfigTopic.checkpoints`, the parsing models, and `CheckpointsConfigProvider`, plus 26 unit tests.
- Established the served schema by locating the proposal doc; confirmed no `checkpoints` topic exists in khepri (`origin/main`, no branch, no open PR), so no backend implementation to conform to yet.
- Empirically verified `JSONDecoder` key conversion with a standalone script rather than assuming: `.convertFromSnakeCase` renames keys read through a keyed container, but not `[String: AnyDecodable]` dictionary keys at any depth. This corrected a wrong field-name assumption that a wire-level test had caught.
- Verified via `RemoteConfigManager.extractInlineBlobs` and khepri's `topics/workflows/publish.py` that an item's payload is always a blob (khepri's `upsert_json_item(content=)` becomes the blob; `item_metadata=` becomes the SDK's `ConfigItem.content`).
- Drafted the review comment posted on purchases-android#3895.

## Human Decisions

- Chose "async parse + sync cached read" for the provider API, and later accepted removing the cache entirely after Toni's reply.
- Chose to support both inline and blob sources, then accepted dropping the inline path once it was shown to be unreachable.
- Confirmed the topic wire name is `checkpoints`, overriding an incorrect reading of the proposal's `checkpoint_rules` heading.
- Directed verification against the workflows topic before acting on Toni's feedback.
- Wrote and approved the exact text of the review comment posted to the Android PR.

## Key Implementation Decisions

- Decision: read the payload via `blobData(for:itemKey:)` only, with no topic-index scan.
  - Rationale: items are keyed by checkpoint name, so the caller's identifier is the item key. `blobData` already resolves inlined vs downloaded blobs and invalidates on identity change.
  - Rejected: pre-parsing the whole topic behind a `GenerationGuardedCache`. Built first, then removed. Its justification was the spec's "evaluation has to be synchronous", but the public checkpoint API is async and a resolved blob read is local, so the constraint did not hold and the cache duplicated manager behavior.
  - Rejected: reading rules from the item's inline `content`. Unreachable, see Agent Contribution.
- Decision: hand-parse from `AnyDecodable` instead of `Decodable` structs.
  - Rationale: a `Decodable` struct throws on the first bad rule and fails the whole payload; per-rule skipping keeps the rest usable while the schema is unstable.
  - Rejected: `Decodable` + `convertFromSnakeCase`, which would be less code. Flagged to the human as a live tradeoff.
- Decision: preserve served rule order, no sorting.
  - Rationale: the proposal states rules arrive in priority order and carry no `order` field.
- Decision: drop a `schedule` whose dates cannot be parsed, keeping the rule.
  - Rationale: an open-ended window would run a rule outside the dates it was scheduled for.
- Decision: do not model `first_seen` / `created_via` / `ordered_rule_ids`.
  - Rationale: dashboard bookkeeping; the SDK reports checkpoint sightings rather than consuming them.

## Files / Symbols Touched

- `Sources/Networking/RemoteConfigTopic.swift`
  - Why: open the read facade for the new topic.
  - Symbols: `RemoteConfigTopic.checkpoints`
  - Review relevance: wire name must match what the backend publishes; a mismatch fails silently.
- `Sources/Networking/CheckpointRules.swift`
  - Why: parsing models.
  - Symbols: `CheckpointRuleSet`, `CheckpointRule`, `CheckpointFrequencyCap`, `CheckpointRuleSchedule`, `CheckpointRuleSet.parse(identifier:blob:)`, `parse(identifier:fields:)`
  - Review relevance: which fields are required vs optional; that rule fields are read at wire spelling (`workflow_id`, `frequency_cap`).
- `Sources/Networking/CheckpointsConfigProvider.swift`
  - Why: topic front door.
  - Symbols: `CheckpointsConfigProviderType`, `CheckpointsConfigProvider.getCheckpoint(_:)`
  - Review relevance: the conflated `nil` cases.
- `Sources/Logging/Strings/RemoteConfigStrings.swift`
  - Why: warn when a rule is skipped.
  - Symbols: `RemoteConfigStrings.checkpointRuleSkipped(checkpoint:reason:)`
- `Tests/UnitTests/Networking/CheckpointRulesTests.swift`, `Tests/UnitTests/Networking/CheckpointsConfigProviderTests.swift`
  - Why: parsing and end-to-end read coverage.
  - Review relevance: `testKeepsPayloadKeysVerbatimWhileItemMetadataIsCamelCased` pins the key-casing contract.

## Dependencies / Config / Migrations

- None. All types are internal, so no public API surface change and no `api/*.swiftinterface` update.

## Validation

- Commands run:
  - `swift build --target RevenueCat`: clean.
  - `xcodebuild test -workspace RevenueCat-Tuist.xcworkspace -scheme UnitTests … -only-testing:UnitTests/CheckpointRulesTests -only-testing:UnitTests/CheckpointsConfigProviderTests`: 26 tests, 0 failures.
  - Same command for `RemoteConfigManagerTests`, `WorkflowsConfigProviderTests`, `RemoteConfigSourceProviderTests`, `UiConfigProviderTests`, `RemoteConfigurationDecodingTests`, `RemoteConfigIntegrationTests`: 258 tests, 0 failures.
  - `swiftlint lint` on all changed files: no violations.
- Manual verification:
  - No device or simulator app run: nothing is wired into `Purchases`, so there is no runtime path to exercise.
  - No real backend response inspected: the topic is not served yet.
- CI:
  - Not captured at time of writing.

## Validation Gaps

- Never exercised against a real `/v1/config` response, since the backend does not serve this topic yet. Tests encode the proposal's example payload, not observed traffic.
- No test asserts behavior when remote config is disabled, beyond the `nil` return that path shares with other cases.
- Only iOS unit tests run; other platform test lanes not run locally.

## Review Focus

- Is the topic wire name `checkpoints` correct? A wrong name returns `nil` forever with no error.
- Should the payload decode into `Decodable` structs instead of hand-parsed `AnyDecodable`, accepting all-or-nothing failure for less code?
- Is `workflow_id` the right sole required field, or should a rule without an `audience` also be rejected?
- Is dropping an unparseable `schedule` while keeping the rule the behavior we want, or should the rule be dropped too?
- Should `frequency_cap.type` stay a raw `String` rather than a typed value while the vocabulary is unsettled?

## Risks / Reviewer Notes

- Risk: the schema shifts before V1 and parsing silently reads `nil` for renamed fields.
  - Evidence: the proposal's own callout says the schema may shift; it is only used internally for testing.
  - Mitigation: unknown fields ignored, malformed rules skipped and logged; `CheckpointRulesTests` encodes the documented example so a rename shows as a test failure.
- Risk: key-casing confusion, since metadata keys are camelCased and payload keys are not.
  - Evidence: verified empirically; an earlier revision read the wrong field names and a wire-level test caught it.
  - Mitigation: documented on `parse(identifier:fields:)` and pinned by a test asserting both sides.
- Risk: `nil` conflates "unconfigured", "unresolvable" and "disabled".
  - Evidence: `blobData` returns `nil` for all three.
  - Mitigation: none yet; noted as a follow-up for when `noMatch` vs `configurationUnavailable` is surfaced.
- Risk: divergence from Android if only one platform's read path changes.
  - Evidence: both now use `getCheckpoint(identifier)` over `blobData`; the Android model is still an empty placeholder.
  - Mitigation: comment posted on purchases-android#3895; naming to be agreed when that model is filled in.

## Non-goals / Out of Scope

- Rule evaluation, audience matching, frequency capping, schedule-window checks.
- Wiring into `Purchases` or any consumer.
- The public `checkpoint()` API (#7345) and `RulesEngine` integration.
- Checkpoint sighting/analytics reporting.

## Omitted Context

- Raw transcript, unrelated exploration, sensitive details, repetitive attempts, and chain-of-thought-style content were omitted.

</details>

---

> [!NOTE]
> **Low Risk**
> Internal networking-only ingestion with no Purchases wiring or public API changes; failures degrade to nil and skipped rules rather than crashing callers.
> 
> **Overview**
> Adds ingestion of the **`checkpoint_rules`** remote-config topic so checkpoint rule payloads can be read and decoded before any evaluation exists. **`RemoteConfigTopic.checkpointRules`** (`wireName: `checkpoint_rules``) is registered alongside existing topics.
> 
> **`CheckpointRuleSet`** and **`CheckpointRule`** decode blob JSON with **`Decodable`**, preserving served rule order and requiring **`audience_id`** and **`workflow_id`**. Malformed or incomplete rules are skipped (with **`checkpointRuleSkipped`** logging) so one bad entry does not fail the whole checkpoint.
> 
> **`CheckpointsConfigProvider`** mirrors **`WorkflowsConfigProvider`**: **`getCheckpoint(_:)`** loads data via **`RemoteConfigManager.blobData(for: .checkpointRules, itemKey:)`** and returns **`nil`** when the topic is absent, the item is missing, or decoding fails. Unit tests cover decoding edge cases and end-to-end reads through a real **`RemoteConfigManager`** with faked transport/storage.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 7892f185d16ceae2b9614e806a48d40025e2ab27. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: Sources/Networking/CheckpointRules.swift [hunk 1]
```diff
@@ -0,0 +1,223 @@
+//
```
Comment (facumenzella) [inline Sources/Networking/CheckpointRules.swift:1]: I need to prune some comments

File: Sources/Networking/CheckpointRules.swift [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+        return Self.parse(identifier: identifier, fields: fields)
+    }
+
+    /// Parses one checkpoint's id and rules from its decoded payload. A malformed rule is skipped rather than
+    /// failing the checkpoint, and unknown fields are ignored so the backend can extend the schema without a
+    /// client release.
+    ///
+    /// Keys are read at their wire spelling (`workflow_id`, `frequency_cap`): the payload is decoded as a
+    /// dictionary, and `.convertFromSnakeCase` only renames keys read through a keyed container. That holds at
+    /// every depth here, unlike an item's `content`, whose own keys `ConfigItem.init(from:)` does camelCase.
```
Comment (facumenzella) [inline Sources/Networking/CheckpointRules.swift]: too much

File: Sources/Networking/CheckpointRules.swift [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+    /// Dropped rather than kept as open-ended, which would run the rule outside its dates.
+    private static func parseSchedule(_ field: AnyDecodable?) -> CheckpointRuleSchedule? {
+        guard case let .object(fields)? = field else { return nil }
+
+        let start = Self.parseDate(fields[Self.startKey])
+        let end = Self.parseDate(fields[Self.endKey])
+        guard start != nil || end != nil else { return nil }
+
+        return CheckpointRuleSchedule(start: start, end: end)
+    }
```
[the remark this answers is not in the document]
Comment (vegaro) [reply Sources/Networking/CheckpointRules.swift]: I think this is correct. The behavior is not matching the comment "`/// Dropped rather than kept as open-ended, which would run the rule outside its date`

File: Sources/Networking/RemoteConfigTopic.swift [hunk 4]
```diff
@@ -16,6 +16,7 @@ enum RemoteConfigTopic: String {
     case workflows
     case uiConfig = "ui_config"
     case sources
+    case checkpoints
```
Comment (vegaro) [inline Sources/Networking/RemoteConfigTopic.swift]: I think the name the backend is sending is `checkpoint_rules`

File: Sources/Networking/CheckpointRules.swift [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+        self.id = id
+        self.rules = rules
+    }
+
+}
+
+struct CheckpointRule: Equatable, Sendable {
+
+    let id: String?
+    /// Reference to the targeted audience, whose predicate arrives in a separate topic. `nil` targets everyone.
```
Comment (vegaro) [inline Sources/Networking/CheckpointRules.swift]: `` `nil` targets everyone. ``<br><br>should we be doing this? or not target anyone to avoid unintended targetting

File: Sources/Networking/CheckpointRules.swift [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+        self.audienceId = audienceId
+        self.workflowId = workflowId
+        self.frequencyCap = frequencyCap
+        self.schedule = schedule
+    }
+
+}
+
+/// `type` stays the raw wire value: frequency capping is out of V1 scope and its vocabulary isn't settled.
+struct CheckpointFrequencyCap: Equatable, Sendable {
```
Comment (rickvdl) [inline Sources/Networking/CheckpointRules.swift]: Should we introduce this if we're not going to use it? Just thinking about possible ways parsing may fail down the line once we start sending this.

File: Sources/Networking/CheckpointRules.swift [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+    }
+
+}
+
+// MARK: - Parsing
+
+extension CheckpointRuleSet {
+
+    /// `nil` when the bytes aren't a JSON object.
+    static func parse(identifier: String, blob data: Data) -> CheckpointRuleSet? {
```
Comment (rickvdl) [inline Sources/Networking/CheckpointRules.swift]: Nit: not against having this, but the topic manager should have a helper for this, this should work whenever we want to parse it, if we need more customization we should be able to do that in the Codable implementation
```
try await manager.blobData(
    for: .checkpoints,
    itemKey: identifier,
    as: CheckpointRuleSet.self
)
```

File: Sources/Networking/CheckpointRules.swift (hunk 6 above)
Comment (vegaro) [reply Sources/Networking/CheckpointRules.swift]: I agree, let's not shoot ourselves in the foot

File: Sources/Networking/RemoteConfigTopic.swift (hunk 4 above)
Comment (facumenzella) [reply Sources/Networking/RemoteConfigTopic.swift]: I had that, but I think I changed it based on the android pr 🙃

File: Sources/Networking/CheckpointRules.swift (hunk 5 above)
Comment (facumenzella) [reply Sources/Networking/CheckpointRules.swift]: fair, flipping it

File: Sources/Networking/CheckpointRules.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (facumenzella) [reply Sources/Networking/CheckpointRules.swift]: removed the schedule entirely in the end, it's not V1, so this one goes away with it. when we do bring it back for evaluation it needs to fail closed, since khepri publishes rules in `scheduled` state and not just `active`

File: Sources/Networking/CheckpointRules.swift (hunk 6 above)
Comment (facumenzella) [reply Sources/Networking/CheckpointRules.swift]: gone. did the same with `schedule` for the same reason, so a rule is now just `id`, `audience` and `workflow_id`

File: Sources/Networking/CheckpointRules.swift (hunk 7 above)
Comment (facumenzella) [reply Sources/Networking/CheckpointRules.swift]: done, it's `blobData(for:itemKey:as:)` now with the models `Decodable`. kept the per-rule tolerance in a small `SkippableRule` wrapper so one bad rule doesn't take out the whole checkpoint

side effect I didn't expect: `JSONDecoder.default` throws on a bad date, which is what fixed the schedule thread above before I deleted the field anyway

File: Sources/Networking/CheckpointRules.swift [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+extension CheckpointRule: Decodable {
+
+    private enum CodingKeys: String, CodingKey {
+        case id
+        case audienceId = "audience"
+        case workflowId
+        case schedule
+    }
+
+}
```
[the remark this answers is not in the document]
Comment (facumenzella) [reply Sources/Networking/CheckpointRules.swift:84]: right, the synthesized `Decodable` lost that. added a custom `init(from:)` that throws on an empty `audience` or `workflow_id`, plus the test back

File: Sources/Networking/RemoteConfigTopic.swift [hunk 9]
```diff
@@ -16,6 +16,7 @@ enum RemoteConfigTopic: String {
     case workflows
     case uiConfig = "ui_config"
     case sources
+    case checkpointRules = "checkpoint_rules"
```
[the remark this answers is not in the document]
Comment (facumenzella) [reply Sources/Networking/RemoteConfigTopic.swift:19]: `checkpoint_rules` is the right one, khepri registers `TopicName('checkpoint_rules')` and Android uses the same wire name. the provider doc was the stale bit, fixed

the tests point is fair though, they all derive the key from the enum so they'd never catch a bad name. added one that pins the literal

**Review by rickvdl — APPROVED** [review]
I think this looks good!

File: Sources/Networking/CheckpointRules.swift [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+            .compactMap(\.rule)
+    }
+
+}
+
+extension CheckpointRule: Decodable {
+
+    private enum CodingKeys: String, CodingKey {
+        case id
+        case audienceId = "audience"
```
Comment (vegaro) [inline Sources/Networking/CheckpointRules.swift]: we are updating this to be `audienceId` in the backend

File: Sources/Networking/CheckpointRules.swift [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+//  Created by Facundo Menzella.
+//  Copyright © 2026 RevenueCat, Inc. All rights reserved.
+
+import Foundation
+
+/// One checkpoint's rules, which decide whether it resolves to a workflow. Parsed only, not evaluated.
+struct CheckpointRuleSet: Equatable, Sendable {
+
+    /// The checkpoint's backend id, as opposed to its name, which is the topic item key.
+    let id: String?
```
Comment (vegaro) [inline Sources/Networking/CheckpointRules.swift]: I have removed this one as well in Android since we won't use it yet and find it confusing with the "identifier"

**Review by vegaro — APPROVED** [review]
